utils/audiotool.1
.\" audiotool - manipulation telephony audio files.
.\" Copyright (c) 2009-2010 David Sugar <dyfet@gnutelephony.org>
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc.,59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH audiotool "1" "January 2010" "GNU ccAudio" "GNU Telephony"
.SH NAME
audiotool \- manipulate telephony audio files.
.SH SYNOPSIS
.B audiotool \fI[options]\fR \fIcommand\fR \fIfiles...\fR
.br
.SH DESCRIPTION
The audiotool command is used to examine and manipulate telephony audio files
such as those which might be used in GNU Bayonne.  It can be used to convert
audio to different formats or join audio files together as well.  Current
releases of audiotool do not support resampling, and are principally focused
on either raw audio samples or SUN .au contained audio.  In the future .wav
contained audio files may also be supported.
.SH OPTIONS
.TP
\fB--buffer\fR \fImsecs\fR
Specify ``size'' of user space buffering in milliseconds.  Buffering is used
to reduce how frequently user space audio data has to be pushed into kernel
space file operations.  Because some codecs and conversions have fixed
and specific frame sizes, you want to choose a value that will be a good
multiplier of any possible codec that might be used.  The default is 120
milliseconds, which is good for the list of suggested framing values.  This
also allows for predictive I/O scheduling.
.TP
\fB--codecs\fR
Lists codec channel encoding formats that are supported.
.TP
\fB--encoding\fR \fIformat\fR
Specify the encoding format that audiotool will use for internal operations.
Files opened in different encoding formats will be transcoded to this format
if possible.  The default is pcmu.  The list of supported formats is shown
with the \fI-codecs\fR option.
.TP
\fB--framing\fR \fImsecs\fR
Specify audio framing in milliseconds.  Generally you
should use values like 10, 15, 20, 30, 50, or 60 milliseconds.  If no
framing is used, then the default requirements of the codec will be
selected, and this is often 20 milliseconds.  Some commands are only
supported if ulaw or linear is chosen.
.TP
\fB--help\fR
Outputs help screen for the user.
.TP
\fB--note\fR \fI``text''\fR
Specify a file annotation when creating a new file.
.TP
\fB--output\fR \fIformat\fR
Specify the output encoding that audiotool wil use for creating a new file
such as for the \fBcreate\fR command.  This is useful when it cannot be
determined from the file extension alone, such as when one wants to use a
specific codec format for creating a .au file.
.TP
\fB--verbose\fR
Display extra debugging information for some commands.
.TP
\fB--version\fR
Outputs package version and copyright notice.
.SH COMMANDS
.TP
\fBappend\fR \fInewfile\fR \fIfiles...\fR
Append to existing audio file, converting other files in the process.
.TP
\fBcreate\fR \fInewfile\fR \fIfiles...\fR
Create a new audio file, converting existing files in the process.  The 
encoding for the target file might be specified by the file extension, or 
optionally through the \fB-output\fR option.
.TP
\fBinfo\fR \fIfiles...\fR
Output core library information and (when verbose) buffering stats for 
each of the files listed.  When used verbose, this can be useful to better 
understand how buffering factors and framing works as well.
.TP
\fBpipe\fR \fIfiles...\fR
Pipe the raw output of the specified files to stdout based on the channel
encoding format.  The encoding format is by default 8khz ulaw, and hence the
output can be directly piped to and played on a oss /dev/audio or sun 
audio device if the format is left unchanged.
.TP
\fBtext\fR \fIfile\fR
Output the annotation stored in a SUN .au (and later .wav) file.
.TP
\fBverify\fR \fIfile\fR
Tests if audio file is readable.  If used verbose, it will report frames 
read successfully vs I/O context switches to kernel space.
.SH "EXIT STATUS"
Normally exit status is 0 if the request was successfully processed by the
sipwitch service daemon.  An exit status of 1 indicates a file referenced
does not exist or cannot be accessed.  A 2 indicates that the encoding format
of the specified file is unsupported or cannot be converted.  An exit status
of 3 indicates an I/O failure happened. Any command syntax error will return 
a 4.
.SH AUTHOR
.B audiotool
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to sipwitch-devel@gnu.org.
.SH COPYRIGHT
Copyright \(co 2009 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.




NEWS
uCommon Library NEWS -- history of visible user changes. Updated 2008-04-15.

uCommon C++ 0.4 codebase completed and ready for release with the introduction
of the GNU GPL V3.  This is a complete rewrite of UCommon as a C++ library for
threading and design patterns suitable for deeply embedded targets.

uCommon C++ 0.7 codebase now has core documentation and basic library theory in
header files.  This documentation can be generated with doxygen or by using
"make doxy".

uCommon C++ 1.0 release has been simplified, fully documented, and modified in
specific ways to simplify the migration and merge with GNU Common C++ to form
GNU Common C++ 2.0.

uCommon C++ 1.5 release begins the process of migrating GNU Common C++ into the
core GNU uCommon C++ codebase. Some GNU Common C++ classes, including tcp
streaming, xml parser, and persist, have been revised and integrated for the
1.5.0 release itself.  Other GNU Common C++ classes will be integrated as
requested and needed.  The 1.5/1.x series will actually be the last UCommon
release sets, as GNU uCommon C++ 2.0 will instead be released as GNU Common C++
2 2.0.

uCommon C++ 2.0 released under the GNU Lesser General Public License, Version 3
or later.

uCommon C++ 2.1 is considered a refactored abi release offering greater
clarity, consistency of use, and additional features that were migrated from
Common C++. The 2.1 release series may continue to introduce additional
classes.

ucommon 2.2 merges other ucommon related libraries to create a single ucommon
framework for application development.



utils/audiotool.cpp
// Copyright (C) 1999-2005 Open Source Telecom Corporation.
// Copyright (C) 2006-2009 David Sugar, Tycho Softworks.
//
// This file is part of GNU uCommon C++.
//
// GNU uCommon C++ is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published
// by the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// GNU uCommon C++ is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with GNU uCommon C++.  If not, see <http://www.gnu.org/licenses/>.

#include <ucommon/audio.h>
#include <cstdio>
#include <config.h>

static const  char *delfile = NULL;

using namespace UCOMMON_NAMESPACE;

static const char *note = NULL;
static const audiocodec *oformat = NULL;
static const audiocodec *encoding = NULL;
static audiobuffer *input = NULL, *output = NULL;
static timeout_t framing = 0;
static timeout_t buffering = 120;

static shell::flagopt helpflag('h', "--help", _TEXT("display this list"));
static shell::flagopt althelp('?', NULL, NULL);
static shell::numericopt bufopt('b', "--buffering", _TEXT("audio buffering (120)"), "msec", 120);
static shell::flagopt codecflag('c', "--codecs", _TEXT("list codecs"));
static shell::stringopt encopt('e', "--encoding", _TEXT("channel encoding (pcmu)"), "codec", "pcmu");
static shell::numericopt frameopt('f', "--framing", _TEXT("audio framing (20)"), "msec", 20); 
static shell::stringopt noteopt('n', "--note", _TEXT("text annotation for new files"), "\"text\"", "");
static shell::stringopt outopt('o', "--output", _TEXT("encoding for new files (pcmu)"), "codec", "pcmu");
static shell::flagopt verbose('v', "--verbose", _TEXT("display extra debugging info"));
static shell::flagopt version(0, "--version", _TEXT("display software version"));

extern "C" {
#ifdef  _MSWINDOWS_
	static BOOL WINAPI down(DWORD ctrltype)
	{
		if(delfile) {
			remove(delfile);
			delfile = NULL;
		}
		exit(ctrltype);
		return TRUE;
	}
#else
	static void down(int signo)
	{
		if(delfile)
			remove(delfile);
		exit(signo);
	}
#endif
}

extern "C" {
	static void stop(void)
	{
		if(input)
			audio::release(input);
		if(output)
			audio::release(output);
		input = output = NULL;
	}
}

static void codecs(void)
{
	linked_pointer<audiocodec> ac = audiocodec::first();

	while(is(ac)) {
		printf("%s %ldmsec\n", ac->getName(), ac->getFraming());
		ac.next();
	}
	exit(0);
}

static void usage(void)
{
	printf("%s\n", _TEXT("Usage: audiotool [options] command files...\n"
		"Options:\n"));
	shell::help();
	printf("%s", _TEXT("Commands:\n"
		"  append file files           Append audio to existing file\n"
		"  create newfile files        Create new file from existing ones\n"
		"  info files                  Display info for specified audio files\n"
#ifndef	_MSWINDOWS_
		"  pipe files                  Pipe raw transcoded audio to stdout\n"
#endif
		"  text file                   Display annotation for audio file if set\n"
		"  verify file                 Verify readability of an audio file\n"
	));
	printf("%s\n", _TEXT("Report bugs to sipwitch-devel@gnu.org"));
    exit(0);
}
	
static void versioninfo(void)
{
	printf("audiotool " VERSION "\n"
		"Copyright (C) 2009-2010 David Sugar, Tycho Softworks\n"
		"This is free software; see the source for copying conditions.  There is NO\n"
		"warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n");
    exit(0);
}

static void text(const char *fn)
{
	char buf[256];

	audiofile::info(input, fn, framing);
	switch(input->status()) {
	case audio::INVFILE:
		fprintf(stderr, "*** %s: missing or inaccessible\n", fn);
		exit(1);
	case audio::INVALID:
		fprintf(stderr, "*** %s: invalid or unsupported encoding\n", fn);
		exit(2);
	default:
		break;
	}
	input->text(buf, sizeof(buf));
	if(buf[0])
		printf("%s\n", buf);
	exit(0);
}

static void info(const char *fn)
{
	char buf[256];
	const audiocodec *encoding;
	const char *cp;

	audiofile::info(input, fn, framing);
	switch(input->status()) {
	case audio::INVFILE:
		fprintf(stderr, "*** %s: missing or inaccessible\n", fn);
		exit(1);
	case audio::INVALID:
		fprintf(stderr, "*** %s: invalid or unsupported encoding\n", fn);
		exit(2);
	default:
		break;
	}

	encoding = input->getEncoding();
	cp = strrchr(fn, '/');
	if(!cp)
		cp = strrchr(fn, '\\');
	if(!cp)
		cp = strrchr(fn, ':');
	if(cp)
		++cp;
	else
		cp = fn;

	unsigned used = input->getBuffering() + input->getBackstore() + input->getConversion();
	printf("audiofile %s\n", cp);

	if(is(verbose)) {
		printf("available %u\n", input->getAvailable());
		printf("buffering %u\n", input->getBuffering());
		printf("backstore %u\n", input->getBackstore());
		printf("converter %u\n", input->getConversion());
		printf("underused %u\n", input->getAvailable() - used);
		printf("framesize %u\n", input->framesize());
	}

	printf("encoding  %s\n", encoding->getName());
	printf("framing   %u msec\n", (unsigned)encoding->getFraming());
		input->text(buf, sizeof(buf));

	printf("length    %02lu:%02lu:%02lu.%03lu\n", 
		(input->length() /3600000l),
		(input->length() / 60000l) % 60l,
		(input->length() / 1000l) % 60l, 
		input->length() % 1000l);

	if(buf[0])
		printf("label     %s\n", buf);
}

static void append(const char *to, char **argv)
{
	unsigned count = 0, len;
	audio::encoded_t data;
	const char *from;

	audiofile::append(output, to, framing);
	switch(output->status()) {
	case audio::INVFILE:
		fprintf(stderr, "*** %s: cannot access\n", to);
		exit(1);
	case audio::INVALID:
		fprintf(stderr, "*** %s: invalid or unsupported encoding\n", to);
		exit(2);
	default:
		break;
	}

	while(*argv) {
		from = *(argv++);

		audiofile::access(input, from, framing);
		switch(input->status()) {
		case audio::INVFILE:
			fprintf(stderr, "*** %s: missing or inaccessible\n", from);
			exit(1);
		case audio::INVALID:
			fprintf(stderr, "*** %s: invalid or unsupported encoding\n", from);
			exit(2);
		default:
			break;
		}

		while(NULL != (data = input->get())) {
			len = output->put(data);
			if(!len) {
				fprintf(stderr, "*** %s: output failed\n", to);
				exit(3);
			}
			++count;
		}

		output->release();
		switch(input->status()) {
		case audio::END:
			break;
		default:
			fprintf(stderr, "*** %s: input failed\n", from);
			exit(3);
		}
	}

	if(is(verbose)) {
		printf("converted %u frames\n", count);
		printf("reading %lu contexts\n", input->getContexts());
		printf("writing %lu contexts\n", output->getContexts());
	}

	exit(0);
}

static void create(const char *to, char **argv)
{
	unsigned count = 0, len;
	audio::encoded_t data;
	const char *from;

	::remove(to);
	audiofile::create(output, to, framing, note, oformat);
	switch(output->status()) {
	case audio::INVFILE:
		fprintf(stderr, "*** %s: cannot create\n", to);
		exit(1);
	case audio::INVALID:
		fprintf(stderr, "*** %s: invalid or unsupported encoding\n", to);
		exit(2);
	default:
		break;
	}

	while(*argv) {
		from = *(argv++);

		audiofile::access(input, from, framing);
		switch(input->status()) {
		case audio::INVFILE:
			fprintf(stderr, "*** %s: missing or inaccessible\n", from);
			exit(1);
		case audio::INVALID:
			fprintf(stderr, "*** %s: invalid or unsupported encoding\n", from);
			exit(2);
		default:
			break;
		}

		while(NULL != (data = input->get())) {
			len = output->put(data);
			if(!len) {
				fprintf(stderr, "*** %s: output failed\n", to);
				exit(3);
			}
			++count;
		}
		output->release();

		switch(input->status()) {
		case audio::END:
			break;
		default:
			fprintf(stderr, "*** %s: input failed\n", from);
			exit(3);
		}
	}

	if(is(verbose)) {
		printf("converted %u frames\n", count);
		printf("reading %lu contexts\n", input->getContexts());
		printf("writing %lu contexts\n", output->getContexts());
	}

	exit(0);
}

#ifndef	_MSWINDOWS_
static void pipe(const char *fn)
{
	audio::encoded_t data;
	size_t len = 0;

	audiofile::access(input, fn, framing);
	switch(input->status()) {
	case audio::INVFILE:
		fprintf(stderr, "*** %s: missing or inaccessible\n", fn);
		exit(1);
	case audio::INVALID:
		fprintf(stderr, "*** %s: invalid or unsupported encoding\n", fn);
		exit(2);
	default:
		break;
	}

	while(NULL != (data = input->get())) {
		len = input->framesize(data);
		if(len) {
			if(::write(1, data, len) < (ssize_t)len) {
				fprintf(stderr, "*** %s: output failed\n", fn);
				exit(3);
			}
		}		
	}

	switch(input->status()) {
	case audio::END:
		return;
	default:
		fprintf(stderr, "*** %s: input failed\n", fn);
		exit(3);
	}
}
#endif

static void verify(const char *fn)
{
	unsigned long count = 0;

	audiofile::access(input, fn, framing);
	switch(input->status()) {
	case audio::INVFILE:
		fprintf(stderr, "*** %s: missing or inaccessible\n", fn);
		exit(1);
	case audio::INVALID:
		fprintf(stderr, "*** %s: invalid or unsupported encoding\n", fn);
		exit(2);
	default:
		break;
	}

	while(input->get()) {
		++count;
	}

	if(is(verbose)) {
		printf("read %lu frames\n", count);
		printf("used %lu contexts\n", input->getContexts());
	}

	switch(input->status()) {
	case audio::END:
		exit(0);
	default:
		exit(3);
	}
}

int main(int argc, char **argv)
{
	shell::bind("audiotool");

	char *cp = NULL;
	shell args(argc, argv);

	audio::init();

	if(is(helpflag) || is(althelp))
		usage();

	if(is(version))
		versioninfo();

	if(is(codecflag))
		codecs();

	if(is(bufopt)) {
		buffering = (timeout_t) *bufopt;
		if(buffering < 60 || buffering % 10)
			shell::errexit(1, "*** audiotool: %s: %s\n", "--buffering", _TEXT("must be >= 60 msec and multiple of 10msec"));
	}

	if(is(frameopt)) {
		framing = (timeout_t) *frameopt;
		if(framing % 10 || framing < 10)
			shell::errexit(1, "*** audiotool: %s: %s\n", "--framing",  _TEXT("must be multiple of 10 msec"));
	}

	if(is(encopt)) {
		encoding = audiocodec::find(*encopt);
		if(!encoding)
			shell::errexit(1, _TEXT("*** audiotool: --encoding: %s: unknown encoding\n"), *encopt);
	}

	if(is(outopt)) {
		oformat = audiocodec::find(*outopt);
		if(!oformat)
			shell::errexit(1, _TEXT("*** audiotool: --output: %s: unknown encoding\n"), *outopt);
	}

	if(is(noteopt))
		note = *noteopt;

	atexit(&stop);
	if(!encoding)
		encoding = audiocodec::find("pcmu");

	input = audiobuffer::create(encoding, buffering);
	output = audiobuffer::create(encoding, buffering);
	
#ifdef  _MSWINDOWS_
    SetConsoleCtrlHandler(down, TRUE);
#else
    signal(SIGINT, down);
    signal(SIGTERM, down);
    signal(SIGQUIT, down);
    signal(SIGABRT, down);
#endif	

	argv = args.argv();

	if(argv && *argv)
		cp = *argv;
	else
		usage();

	// check commands...

	while(*cp == '-')
		++cp;

	if(eq(cp, "create")) {
		++argv;
		if(NULL == *argv || NULL == argv[1])
			goto nofiles;
		cp = *(argv++);
		create(cp, argv);
		goto done;
	}

	if(eq(cp, "append")) {
		++argv;
		if(NULL == *argv || NULL == argv[1])
			goto nofiles;
		cp = *(argv++);
		append(cp, argv);
		goto done;
	}
	
	if(eq(cp, "info")) {
		if(NULL == argv[1])
			goto nofiles;
        while(*++argv) {
            info(*argv);
        }
		goto done;
    }

	if(eq(cp, "text")) {
		if(NULL == argv[1])
			goto nofiles;
		if(argv[2])
			goto toomany;
		text(argv[1]);
	}

	if(eq(cp, "verify")) {
		if(NULL == argv[1])
			goto nofiles;
		if(argv[2])
			goto toomany;
		verify(argv[1]);
	}

#ifndef	_MSWINDOWS_
	if(eq(cp, "pipe")) {
		if(NULL == argv[1])
			goto nofiles;
        while(*++argv) {
            pipe(*argv);
        }
		goto done;
    }
#endif

    fprintf(stderr, _TEXT("*** audiotool: %s: unknown command\n"), cp);
    exit(4);

nofiles:
	fprintf(stderr, _TEXT("*** audiotool: %s: no file(s) to process\n"), cp);
	exit(4);

toomany:
	fprintf(stderr, _TEXT("*** audiotool: %s: too many filenames used\n"), cp);
	exit(4);

done:
	return 0;
}

